src/trace-plugin-interface.js

```diff
@@ -0,0 +1,244 @@
+  }
+  var rootContext = that.agent.createRootSpanData(options.name,
+    incomingTraceContext.traceId,
+    incomingTraceContext.spanId,
+    options.skipFrames ? options.skipFrames + 1 : 1);
+  // If the options object passed in has the setHeader field set,
+  // use it to set trace metadata in an outgoing request.
+  if (is.fn(options.setHeader)) {
+    var outgoingTraceContext = rootContext.traceId + '/' +
+      rootContext.spanId;
+    var outgoingHeaderOptions = (incomingTraceContext.options !== null &&
+      incomingTraceContext.options !== undefined) ?
+      incomingTraceContext.options : constants.TRACE_OPTIONS_TRACE_ENABLED;
+    outgoingTraceContext += (';o=' + outgoingHeaderOptions);
+    options.setHeader('x-cloud-trace-context', outgoingTraceContext);
+  }
+  return new Transaction(that.agent, rootContext);
+};
+
+PluginAPI.prototype.getTransaction = function() {
+  if (cls.getRootContext()) {
+    return new Transaction(this.agent, cls.getRootContext());
+  } else {
+    return null;
+  }
+};
+
+/**
+ * Runs the given function in a root span corresponding to an incoming request,
+ * possibly passing it an object that exposes an interface for adding labels
+ * and closing the span.
+ * @param {object} options An object that specifies options for how the root
+ * span is created and propogated. @see PluginAPI.prototype.createTransaction
+ * @param {function(?Transaction)} fn A function that will be called exactly
+ * once. If the incoming request should be traced, a root span will be created,
+ * and this function will be called with a Transaction object exposing functions
+ * operating on the root span; otherwise, it will be called without any
+ * arguments.
+ * @returns The return value of calling fn.
+ */
+PluginAPI.prototype.runInRootSpan = function(options, fn) {
+  var that = this;
+  return this.agent.namespace.runAndReturn(function() {
+    var oldSkipFrames = options.skipFrames;
+    options.skipFrames = options.skipFrames ? options.skipFrames + 2 : 2;
+    var transaction = that.createTransaction(options);
+    options.skipFrames = oldSkipFrames;
+    return fn(transaction);
+  });
+};
+
+/**
+ * Convenience method which obtains a Transaction object with getTransaction()
+ * and calls its runInChildSpan function on the given arguments. If there is
+ * no current Transaction object, the provided function will be called without
+ * arguments.
+ * @param {object} options An object that specifies options for how the root
+ * span is created and propogated. @see Transaction.prototype.runInChildSpan
+ * @param {function(?Transaction)} fn A function that will be called exactly
+ * once. @see Transaction.prototype.runInChildSpan
+ * @returns The return value of calling fn.
+ */
+PluginAPI.prototype.runInChildSpan = function(options, fn) {
+  var transaction = this.getTransaction();
+  if (transaction) {
+    var oldSkipFrames = options.skipFrames;
+    options.skipFrames = options.skipFrames ? options.skipFrames + 1 : 1;
```

Comment:
nit: it would be better if we didn't permute the user-supplied object, even temporarily.